### PR TITLE
Gemma3

### PR DESCRIPTION
**Pull request by vince62s**
Pushing a cleaner PR with changes from Main only.

It works fine for both Llava-type (pixtral/Mistral-24b) and Gemma3.

However still pending but maybe for later PR:

- clean the image_utils.py tools
- revamp the vision.py VisionEncoder to handle batches properly.

Open questions:
It seems that image masking (required for Gemma3?) is still ok for Llava (sounds reasonable). Hence I left the "decoder_in" arg in self.model.decoder() call from inference.py

I removed the patch from previous PR "mask = None" in VisionEncoder / vision.py (not sure why this was there)

### CODE REVIEW

File: eole/config/models.py [hunk 1]
```diff
@@ -670,11 +691,11 @@ def encoder_decoder_type(cls, data: Any) -> Any:
         if not (isinstance(data, dict)):
             return data
         if "encoder" in data.keys():
-            data["encoder"]["encoder_type"] = "transformer"
+            data["encoder"].encoder_type = "transformer"
         else:
             data["encoder"] = {"encoder_type": "transformer"}
         if "decoder" in data.keys():
-            data["decoder"]["decoder_type"] = "transformer"
+            data["decoder"].decoder_type = "transformer"
```
Comment (francoishernandez) [inline eole/config/models.py:698]: Why was this change needed?

File: eole/modules/multi_headed_attn.py [hunk 2]
```diff
@@ -233,13 +241,17 @@ def _compute_attention(
             and not return_attn
             and query.device.type != "cpu"
         ):
+            # TODO: make this configurable (gemma3)
+            kwargs = {"scale": 256**-0.5} if self.is_decoder else {}
```
Comment (francoishernandez) [inline eole/modules/multi_headed_attn.py:245]: Shouldn't we do something about that?

File: eole/modules/rmsnorm.py [hunk 3]
```diff
@@ -24,7 +24,7 @@ def __init__(self, hidden_size: int, eps: float = 1e-5):
         self.eps = eps
         self.weight = nn.Parameter(torch.ones(hidden_size))
 
-    @torch.compile(dynamic=True)
+    # @torch.compile(dynamic=True)
```
Comment (francoishernandez) [inline eole/modules/rmsnorm.py:27]: Commented for debug or real issue?

File: eole/encoders/vision.py [hunk 4]
```diff
@@ -88,19 +89,33 @@ class VisionEncoder(nn.Module):
     def __init__(self, encoder_config, running_config=None):
         super(VisionEncoder, self).__init__()
         self.encoder_config = encoder_config
-        self.rope = build_rope(encoder_config, mode="2d")
+        if encoder_config.position_encoding_type == PositionEncodingType.Learned:
```
Comment (francoishernandez) [inline eole/encoders/vision.py:92]: Might be worth it splitting in two separate vision encoder classes with all the conditions here. (Would probably also be more flexible to add other vision architectures down the line.)

**Review by francoishernandez — COMMENTED** [review]
Some comments/discussion topics.

Additional notes, can be reviewed in other PR(s):
- some new config fields lack helpstrings;
- `image_utils` is a bit messy, we can probably strip it down at some point.

File: eole/encoders/vision.py (hunk 4 above)
Comment (vince62s) [reply eole/encoders/vision.py:92]: probably when the next encoder comes.

File: eole/modules/multi_headed_attn.py (hunk 2 above)
Comment (vince62s) [reply eole/modules/multi_headed_attn.py:245]: yes but unsure what to do

File: eole/config/models.py (hunk 1 above)
Comment (vince62s) [reply eole/config/models.py:698]: not sure why it popped but for sure data["encoder"] is not a dict anymore I think this should have been changed some time ago.
